v2/slogger/rolling_file_appender/rolling_file_appender.go

```diff
@@ -0,0 +1,461 @@
+// rotated.  Rotated log files will have suffix of the form
+// .YYYY-MM-DDTHH-MM-SS or .YYYY-MM-DDTHH-MM-SS-N (where N is an
+// incrementing serial number used to resolve conflicts) appended to
+// them.  Set maxFileSize to a non-positive number if you wish there
+// to be no limit.  maxRotatedLogs specifies the maximum number of
+// rotated logs allowed before old logs are deleted.  If
+// rotateIfExists is set to true and a log file with the same filename
+// already exists, then the current one will be rotated.  If
+// rotateIfExists is set to false and a log file with the same
+// filename already exists, then the current log file will be appended
+// to.  If a log file with the same filename does not exist, then a
+// new log file is created regardless of the value of rotateIfExists.
+// As RotatingFileAppender might be wrapped by an AsyncAppender, an
+// errHandler can be provided that will be called when an error
+// occurs.  It can set to nil if you do not want to provide one.  The
+// return value headerGenerator, if not nil, is logged at the
+// beginning of every log file.
+//
+// Note that after creating a RollingFileAppender with New(), you will
+// probably want to defer a call to RollingFileAppender's Close() (or
+// at least Flush()).  This ensures that in case of program exit
+// (normal or panicking) that any pending logs are logged.
+func New(filename string, maxFileSize int64, maxRotatedLogs int, rotateIfExists bool, headerGenerator func() []string) (*RollingFileAppender, error) {
+	return NewWithStringWriter(filename, maxFileSize, maxRotatedLogs, rotateIfExists, headerGenerator, nil)
+}
+
+func NewWithStringWriter(filename string, maxFileSize int64, maxRotatedLogs int, rotateIfExists bool, headerGenerator func() []string, stringWriterCallback func(*os.File) slogger.StringWriter) (*RollingFileAppender, error) {
+	if headerGenerator == nil {
+		headerGenerator = func() []string {
+			return []string{}
+		}
+	}
+	if stringWriterCallback == nil {
+		stringWriterCallback = func(f *os.File) slogger.StringWriter {
+			return f
+		}
+	}
+
+	absPath, err := filepath.Abs(filename)
+	if err != nil {
+		return nil, err
+	}
+
+	appender := &RollingFileAppender{
+		MaxFileSize:          maxFileSize,
+		MaxRotatedLogs:       maxRotatedLogs,
+		absPath:              absPath,
+		headerGenerator:      headerGenerator,
+		stringWriterCallback: stringWriterCallback,
+	}
+
+	fileInfo, err := os.Stat(absPath)
+	if err == nil && rotateIfExists { // err == nil means file exists
+		return appender, appender.rotate()
+	} else {
+		// we're either creating a new log file or appending to the current one
+		appender.file, err = os.OpenFile(
+			absPath,
+			os.O_WRONLY|os.O_APPEND|os.O_CREATE,
+			0666,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		if fileInfo != nil {
+			appender.curFileSize = fileInfo.Size()
+		}
+
+		return appender, appender.logHeader()
+	}
+}
+
+func (self *RollingFileAppender) Append(log *slogger.Log) error {
```

Comment:
Dan's pointed out that this method is not thread-safe.  In particular, modifying self.curFileSize and reading it later on are not thread-safe operations.